% Project proposal for Computational Genomics s2017
% By Han Den (Andrew Id: hdeng1) and Luyi Ma (Andrew Id: luyim)
% Topic : Classification of miRNA using deep-learning method

\documentclass[letterpaper, 11pt]{article}

%package
\usepackage[top= 0.5in, left = 0.5in, right= 0.5in, bottom= 0.5in]{geometry}
\usepackage{titlesec}
\usepackage{color}

\titleformat*{\section}{\large\bfseries\rmfamily}
\titleformat*{\subsection}{\large\bfseries}
\titleformat*{\subsubsection}{\large\bfseries}
\titleformat*{\paragraph}{\large\bfseries}
\titleformat*{\subparagraph}{\large\bfseries}
\renewcommand{\labelitemi}{$\bullet$}
\renewcommand{\labelitemii}{\textperiodcentered}


\begin{document}

\begin{center}
{\Large
	\textsc{\textbf Classification of microRNA Using Deep-learning Method}
}

\vspace{0.3cm}

\normalsize Project proposal of Computational Genomics Spring 2017

\vspace{0.5cm}

{\small
	Han Den (Andrew Id: hdeng1)
	
	Luyi Ma (Andrew Id: luyim)
}
\end{center}

%body of the proposal
\section{Project Idea}

%Project idea description
\begin{itemize}
\item
A microRNA (abbreviated miRNA) is a small non-coding RNA molecule (containing about 22 nucleotides) found in plants, animals and some viruses, that functions in RNA silencing and post-transcriptional regulation of gene expression. While the majority of miRNAs are located within the cell, some miRNAs, commonly known as circulating miRNAs or extracellular miRNAs, have also been found in the extracellular environment, including various biological fluids and cell culture media~\cite{Wiki}.
So it's of great value to distinguish miRNAs from the whole genome and make accurate classifications.


The deep learning network will be worked as a classifier and we basically use CNN to make the classification. We will use some natural language processing methods to parse DNA sequence and make classifications based on these data. We choose both miRNA and non-miRNA data to train our model and test accuracy. 

\item 
We plan to use data of miRNA from an open-source database~\cite{miData}.
All the miRNA sequence in this database will be changed into DNA sequence based on the complementary rule and used as our positive input training and testing data. We will use permutation methods to generate our negative input training data and use real data in a sample genome as our negative input testing data. The output will be either 0 or 1, indicating it's miRNA or not.
\end{itemize}

%Software design or algorithm to develop
\section{Software Design}
\begin{itemize}
\item
We plan to use Python as our preferred programming language. We will implement algorithms involved in deep learning and natural language processing in Python. And we also plan to use some deep learning frameworks like Tensorflow to help us in feature extraction. To process data and train the model, we plan to use AWS platform. Finally, we may develop a software to wrap out core scripts implementing our models.
\end{itemize}


%Papers to read
\section{Papers to Read}
\begin{itemize}
\item
	\textbf{List fo papers}
	
	\begin{itemize}
	\item
		DNA Sequence Classification by Convolutional Neural Network~\cite{NLP}.
		
	\item
		Convolutional deep belief networks for scalable unsupervised learning of hierarchical representations~\cite{CNN}.
	\end{itemize}
\end{itemize}


%reference
\begin{thebibliography}{1}
\bibitem{Wiki}
\textit{}https://en.wikipedia.org/wiki/MicroRNA

\bibitem{miData}
\textit{}http://www.mirbase.org/

\bibitem{NLP}
Nguyen N G, Tran V A, Ngo D L, et al. \textit{DNA Sequence Classification by Convolutional Neural Network}[J]. Journal of Biomedical Science and Engineering, 2016, 9(05): 280.

\bibitem{CNN}
Lee H, Grosse R, Ranganath R, et al. \textit{Convolutional deep belief networks for scalable unsupervised learning of hierarchical representations}[C]Proceedings of the 26th annual international conference on machine learning. ACM, 2009: 609-616.
\end{thebibliography}
\end{document}